nav2_costmap_2d/src/costmap_2d_ros.cpp

```diff
@@ -96,8 +94,13 @@ Costmap2DROS::Costmap2DROS(
   declare_parameter("observation_sources", rclcpp::ParameterValue(std::string("")));
   declare_parameter("origin_x", rclcpp::ParameterValue(0.0));
   declare_parameter("origin_y", rclcpp::ParameterValue(0.0));
-  declare_parameter("plugin_names", rclcpp::ParameterValue(plugin_names));
-  declare_parameter("plugin_types", rclcpp::ParameterValue(plugin_types));
+  declare_parameter("plugins", rclcpp::ParameterValue(default_plugins));
+  declare_parameter(
```

Comment:
This is potentially declaring bogus plugins that may or may not exist if this value is overridden. Declaring the default plugins is fine above, but then we should get `plugins` declared above. Then, declare and get the value of their `.plugin` for the case its been overridden with different plugins and names. 

I suppose after you get the `plugins`, you could then check if plugins `==` default_plugins to then declare these, else, declare and get the real ones like you do below. We don't want extra plugin parameters defined and declared if not actually used.

Reply:
Ditto with rest

Reply:
example:
- Declare `plugins` with default vector
- Check if `plugins` are the default vector
- If so, then declare these defaults
- If not, then don't and declare / get the `plugin`  type for the entry in the vector.

Reply:
That stops extra params from popping up that aren't "real"